Họ tên:  Lê Hữu Sang

MSSV: 22120310

# HW0: Làm quen với CUDA

Với các GPU tương đối mới thì để biên dịch chỉ cần dùng câu lệnh: \
`nvcc tên-file.cu -o tên-file-chạy`

Nhưng trên Colab mình thường lấy được GPU khá cũ là Tesla K80 với compute capability (phiên bản phần cứng) là 3.7; để biên dịch đúng với GPU khá cũ này thì bạn cần dùng câu lệnh: \
`nvcc -arch=sm_37 tên-file.cu -o tên-file-chạy` \
Trong đó, 37 chính là compute capability của GPU Tesla K80.

Để phòng trường hợp khi làm bài bạn lấy được GPU có compute capability x.x nhưng khi chấm bài Thầy lại lấy được GPU có compute capability khác x.x, dưới đây mình sẽ có đoạn code Python để tự động lấy 2 con số ứng với compute capability của GPU và lưu vào 2 biến `major` và `minor`:


In [44]:
from numba import cuda
major, minor = cuda.get_current_device().compute_capability
print(f'GPU compute capability: {major}.{minor}')

GPU compute capability: 7.5


Một khi đã chạy đoạn code Python ở trên, để biên dịch thì bạn sẽ dùng câu lệnh: \
`nvcc -arch=sm_{major}{minor} tên-file.cu -o tên-file-chạy`

Dưới đây, khi làm bài thì bạn có thể tùy ý thêm/xóa cell. Đừng xóa mấy cell có chữ của Thầy là được.

## Câu 1

In [45]:
!nvcc -arch=sm_{major}{minor} HW0_P1.cu -o HW0_P1.out

In [46]:
!ls

HW0_P1.cu  HW0_P1.out  HW0_P2.cu  HW0_P2.out  sample_data


In [47]:
!./HW0_P1.out

===== GPU Information =====
GPU card's name: Tesla T4
GPU computation capabilities: 7.5
Maximum number of block dimensions: (1024, 1024, 64)
Maximum number of grid dimensions: (2147483647, 65535, 65535)
Maximum size of GPU memory: 15828320256 bytes (14.74 GB)
Amount of constant memory: 65536 bytes
Amount of shared memory per block: 49152 bytes
Warp size: 32


## Câu 2

In [48]:
!nvcc -arch=sm_{major}{minor} HW0_P2.cu -o HW0_P2.out

In [49]:
!ls

HW0_P1.cu  HW0_P1.out  HW0_P2.cu  HW0_P2.out  sample_data


In [50]:

for N in range(6, 25, 2):
    !./HW0_P2.out {1 << N}


Vector size: 64         | Host:    0.004 ms | Device v1:    0.147 ms | Device v2:    0.042 ms
Vector size: 256        | Host:    0.004 ms | Device v1:    0.136 ms | Device v2:    0.028 ms
Vector size: 1024       | Host:    0.008 ms | Device v1:    0.138 ms | Device v2:    0.028 ms
Vector size: 4096       | Host:    0.027 ms | Device v1:    0.134 ms | Device v2:    0.044 ms
Vector size: 16384      | Host:    0.107 ms | Device v1:    0.128 ms | Device v2:    0.031 ms
Vector size: 65536      | Host:    0.436 ms | Device v1:    0.163 ms | Device v2:    0.042 ms
Vector size: 262144     | Host:    1.528 ms | Device v1:    0.088 ms | Device v2:    0.053 ms
Vector size: 1048576    | Host:    4.884 ms | Device v1:    0.093 ms | Device v2:    0.092 ms
Vector size: 4194304    | Host:   18.823 ms | Device v1:    0.244 ms | Device v2:    0.295 ms
Vector size: 16777216   | Host:   76.674 ms | Device v1:    0.825 ms | Device v2:    1.028 ms


| Vector size | Host time | Device time (version 1) | Device time (version 2) |
|-------------:|-----------:|----------------:|----------------:|
| 64           | 0.004      | 0.147           | 0.042           |
| 256          | 0.004      | 0.136           | 0.028           |
| 1024         | 0.008      | 0.138           | 0.028           |
| 4096         | 0.027      | 0.134           | 0.044           |
| 16384        | 0.107      | 0.128           | 0.031           |
| 65536        | 0.436      | 0.163           | 0.042           |
| 262144       | 1.528      | 0.088           | 0.053           |
| 1048576      | 4.884      | 0.093           | 0.092           |
| 4194304      | 18.823     | 0.244           | 0.295           |
| 16777216     | 76.674     | 0.825           | 1.028           |
